addons/im_livechat/models/im_livechat_channel.py

```diff
@@ -165,7 +162,62 @@ def get_mail_channel(self, anonymous_name, user_id=None, country_id=None):
+
+    def _get_random_operator(self):
+        """ Return a random operator from the available users of the channel that have the lowest number of active* livechats.
+
+        * A livechat is considered 'active' if it has at least one message within the last hour.
+
+        (Some annoying conversions have to be made on the fly because this model holds 'res.users' as available operators
+        and the mail_channel model stores the partner_id of the randomly selected operator)
+
+        Returns:
+            user_id (res.users): a random operator
+        """
+        operators = self._get_available_users()
+        if len(operators) == 0:
+            return False
+
+        self.env.cr.execute("""SELECT COUNT(DISTINCT c.id), c.livechat_operator_id
+            FROM mail_channel c
+            LEFT OUTER JOIN mail_message_mail_channel_rel r ON c.id = r.mail_channel_id
+            LEFT OUTER JOIN mail_message m ON r.mail_message_id = m.id
+            WHERE m.create_date > ((now() at time zone 'UTC') - interval '30 minutes')
+            AND c.channel_type = 'livechat'
+            AND c.livechat_operator_id in %s
+            GROUP BY c.livechat_operator_id
+            ORDER BY COUNT(DISTINCT c.id) asc""", (tuple(operators.mapped('partner_id').ids),))
+        active_channels = self.env.cr.dictfetchall()
+
+        # add operators with 0 active livechat at the beginning of the array because we need to keep it sorted by count
+        active_channel_operator_ids = [active_channel['livechat_operator_id'] for active_channel in active_channels]
+        for operator in operators:
+            if operator.partner_id.id not in active_channel_operator_ids:
+                active_channels.insert(0, {
+                    'livechat_operator_id': operator.partner_id.id,
+                    'count': 0
+                })
+
+        lowest_number_of_conversations = False
+        operators_with_lowest_conversations = []
+        for active_channel in active_channels:
+            if lowest_number_of_conversations is False:
+                # init lowest number of conversations with the first element's count since they are sorted 'ascending'
+                lowest_number_of_conversations = active_channel['count']
+
+            # all operators that share the lowest number of active conversations can be randomly selected
+            if active_channel['count'] == lowest_number_of_conversations:
+                operators_with_lowest_conversations.append(active_channel['livechat_operator_id'])
+            else:
+                break
+
+        if not operators_with_lowest_conversations:
+            return False
+        else:
+            operator_with_lowest_conversations = random.choice(operators_with_lowest_conversations)
+            # convert the selected 'partner_id' to its corresponding res.users
+            return next(operator for operator in operators if operator.partner_id.id == operator_with_lowest_conversations)
```

Comment:
If I may, i would suggest to rewrite this to lessen code size and to avoid unnecessary looping : 
```
        ...
        active_channels = self.env.cr.dictfetchall()

        # If inactive operator(s), return one of them
        active_channel_operator_ids = [active_channel['livechat_operator_id'] for active_channel in active_channels]
        inactive_operators = [operator for operator in operators if operator.partner_id.id not in active_channel_operator_ids]
        if inactive_operators:
            return random.choice(inactive_operators)

        # If no inactive operator, active_channels is always filled in as len(operators) > 0 is True at this point.
        # Get the less busy operators using the active_channels first element (since they are sorted 'ascending')
        smaller_count = active_channels[0]['count']
        less_active_operator = random.choice([
            active_channel['livechat_operator_id'] for active_channel in active_channels
            if active_channel['count'] == smaller_count])
        return next(operator for operator in operators if operator.partner_id.id == less_active_operator)
```

Reply:
More of less copy/pasted your code, much better indeed. Thanks!